user_groups:Sends a notification message to users when someone else add them to user group.

Ujjawal3 opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #23642

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

Notification message sent when `Desdemona` removes and adds a user to user group `hamletcharacters`
![Screenshot from 2023-02-05 00-18-52](https://user-images.githubusercontent.com/99073049/216784448-97859731-1a27-48aa-a55b-feb6c8d62c8f.png)




<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Ujjawal3 commented:
@alya please review this PR.

alya commented:
Thanks! 

1. Let's use a group pill in the notification message.
2. Playing with this makes me realize that we should also send a ""{user_full_name} unsubscribed you from the group @{group_name}." message. Would you be up for adding that to this PR?

Ujjawal3 commented:
@alya I have updated the PR with the changes you suggested and updated the screenshot also. Now message is also sent when a user is unsubscribed from user_group.

alya commented:
Thanks! It looks like the unbsubscription notice is missing the "you"?

Ujjawal3 commented:
@alya I have updated the PR and scrrenshot.

alya commented:
Works for me in manual testing! @sahil839 would you be up for reviewing this one? You can mark it for integration review when it looks good to you, unless you think I should retest it.

sahil839 commented on zerver/tests/test_user_groups.py:
```diff
@@ -352,39 +352,73 @@ def test_update_members_of_user_group(self) -> None:
 
+        # A notification message is sent for adding to user group
+        msg_event = [e for e in events if e["event"]["type"] == "message"]
+        self.assert_length(msg_event, 1)
```
Can we also check the message content probably from the message object like we do for notifications sent on changing stream permissions?

Ujjawal3 replied:
Yes ! Now the test checks message content also.

sahil839 commented on zerver/views/user_groups.py:
```diff
@@ -153,6 +218,13 @@ def remove_members_from_group_backend(
+        recipient_users=user_profiles,
+        user_group_name=user_group.name,
+        sub=False,
+        unsub=True,
```
We can also make these variables mandatory kwarg.

sahil839 reviewed (commented):
Posted some comments.

Ujjawal3 commented:
@sahil839 I have updated the PR and addressed all the comments. Please have a look.

sahil839 reviewed (commented):
Posted a couple of minor comments. Otherwise looks good.

Ujjawal3 commented:
@sahil839 I have done the changes. Please have a look. Thanks.

sahil839 commented on zerver/tests/test_user_groups.py:
```diff
@@ -352,39 +353,82 @@ def test_update_members_of_user_group(self) -> None:
+        # Test user adding itself to user group
+        desdemona = self.example_user("desdemona")
+        self.login_user(desdemona)
+        add = [desdemona.id]
```
Here also we can avoid creating this `add` variable.

sahil839 commented on zerver/tests/test_bots.py:
```diff
@@ -414,6 +416,39 @@ def test_bot_add_subscription(self) -> None:
+            "description": "Support team",
+        }
+
+        self.client_post("/json/user_groups/create", info=params)
```
We can use existing user group of hamletcharacters here, there is no need to create a new group.

sahil839 commented on zerver/tests/test_bots.py:
```diff
@@ -414,6 +416,39 @@ def test_bot_add_subscription(self) -> None:
+        msg_event = [e for e in events_bot if e["event"]["type"] == "message"]
+        self.assert_length(msg_event, 0)
+
+        self.assertEqual(UserGroupMembership.objects.filter(user_group=user_group).count(), 2)
```
Would be good to add a test for checking case of removing bot from group, just for the sake of completion.

Ujjawal3 replied:
Should I write a new test for removing bot from group or should I make changes to the test of adding bot to user group?

sahil839 commented on zerver/tests/test_user_groups.py:
```diff
@@ -352,39 +353,82 @@ def test_update_members_of_user_group(self) -> None:
+        expected_notification = (
+            f"{silent_mention_syntax_for_user(hamlet)} unsubscribed you from the group @_*support*."
+        )
+        self.assertEqual(msg_event[0]["event"]["message_dict"]["content"], expected_notification)
```
Here also would be good to test the case when removing yourself from user group.

sahil839 reviewed (commented):
Thanks! Posted a few more comments. Also please update the commit message to follow the [guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages), specifically the commit summary part and line-wrapping in the commit message.

Ujjawal3 commented:
@sahil839 I have updated the PR as per the changes you suggested. Please have a look.

sahil839 commented on zerver/tests/test_user_groups.py:
```diff
@@ -350,42 +351,98 @@ def test_update_members_of_user_group(self) -> None:
+        expected_notification = (
+            f"{silent_mention_syntax_for_user(hamlet)} subscribed you to the group @_*support*."
+        )
+        self.assertEqual(msg_event[0]["event"]["message_dict"]["content"], expected_notification)
```
Now reading this again, instead of testing the event, we should check the actual `Message` object and its content. You can either use `get_last_message` helper or can create something like `get_topic_messages` that would be used to get PMs with a particular user.

sahil839 commented on zerver/views/user_groups.py:
```diff
@@ -124,6 +176,7 @@ def add_members_to_group_backend(
     user_group = access_user_group_by_id(user_group_id, user_profile)
     user_profiles = user_ids_to_users(members, user_profile.realm)
     existing_member_ids = set(get_direct_memberships_of_users(user_group, user_profiles))
+    acting_user = user_profile
```
There is no need of this extra variable.

Ujjawal3 replied:
Removing this variable causes a bug beacuse the variable `user_profile` is used for iterating inside a loop in the function.
So after iteration is completed `user_profile` is no longer the same person which made request.

If we need to remove this variable then variable used for iterating loop should be changed.

sahil839 replied:
Ok. thanks! We can leave this as it is.

timabbott replied:
I think it's generally better to address this sort of thing by using a different loop variable -- redefining the meaning of a variable midfunction leads to confusing code and that leads to bugs. Something of the form `foo_user` should work great.

Ujjawal3 replied:
Yes, I have done the changes in a separate commit.

sahil839 reviewed (commented):
Posted a couple of comments. Should probably be the final round of review from my side.

Ujjawal3 commented:
@sahil839 I have updated the PR as per your comments. Please have a look.

sahil839 commented:
Thanks! Looks good to me.

timabbott commented on zerver/views/user_groups.py:
```diff
@@ -115,6 +119,54 @@ def update_user_group_backend(
+    user_group_name: str,
+    *,
+    send_subscription_message: bool,
+    send_unsubscription_message: bool,
```
I like the use of mandatory kwargs, but it might be better for both of these to default to false, so that callers only need to pass `send_subscription_message=True` if they want that behavior.

timabbott commented on zerver/views/user_groups.py:
```diff
@@ -115,6 +119,54 @@ def update_user_group_backend(
     return json_success(request, data)
 
 
+def send_messages_for_user_groups(
```
This name could be improved. It's a fairly subtle thing this is doing, so maybe we just need a long name, like `notify_for_user_group_subscription_changes`.

timabbott commented on zerver/views/user_groups.py:
```diff
@@ -115,6 +119,54 @@ def update_user_group_backend(
+def send_messages_for_user_groups(
+    acting_user: UserProfile,
+    recipient_users: List[UserProfile],
+    user_group_name: str,
```
We should take the `UserGroup` object rather than its name as a parameter.

timabbott commented on zerver/views/user_groups.py:
```diff
@@ -115,6 +119,54 @@ def update_user_group_backend(
+    notifications = []
+    sender = get_system_bot(settings.NOTIFICATION_BOT, realm.id)
+    for recipient_user in recipient_users:
+        if recipient_user.email == acting_user.email:
```
Never compare emails -- you should compare the objects' IDs, since those are guaranteed to be unique.

timabbott commented on zerver/views/user_groups.py:
```diff
@@ -115,6 +119,54 @@ def update_user_group_backend(
+            continue
+        if recipient_user.is_bot:
+            # Don't send notification message to bots.
+            continue
```
What happens if someone tries to edit the group membership of a deactivated user? We certainly don't want to send these notifications to deactivated users. I recommend including a deactivated user in your test setup; it'll be a case we want the tests to enforce since it's the kind of thing that tends to get messed up when refactoring.

Ujjawal3 replied:
Earlier I missed the case of deactivated user but now I have done the changes and message will not be sent to deactivated user. Also added a test for this case.

timabbott commented on zerver/views/user_groups.py:
```diff
@@ -115,6 +119,54 @@ def update_user_group_backend(
+    mention_backend = MentionBackend(realm.id)
+
+    notifications = []
+    sender = get_system_bot(settings.NOTIFICATION_BOT, realm.id)
```
I think calling this variable `notification_bot` would be more readable.

timabbott commented on zerver/views/user_groups.py:
```diff
@@ -115,6 +119,54 @@ def update_user_group_backend(
+                )
+            if send_unsubscription_message:
+                message = _(
+                    "{user_full_name} unsubscribed you from the group {group_name}."
```
These should be "added you to / removed you from" -- we're reserving the word "subscribed" for stream membership.

(I know the original issue called for these strings, but I've checked with Alya and that was an oversight).

timabbott commented on zerver/tests/test_bots.py:
```diff
@@ -414,6 +416,35 @@ def test_bot_add_subscription(self) -> None:
+        self.assertEqual(UserGroupMembership.objects.filter(user_group=user_group).count(), 2)
+
+        # No notification message is sent for removing a bot
+        self.assertEqual(self.get_last_message(), initial_last_message)
```
I find this sort of test tends to not be very effective, because it is separated from the non-bot case. 

I think we should delete this test, and just extend `test_update_members_of_user_group` to add/remove some bot users in the same API request as non-bot users, so that we're doing the assertions in a code path that also has notifications being sent to other users. (I think the same idea will work for deactivated users)

See https://zulip.readthedocs.io/en/latest/testing/philosophy.html#share-test-setup-code for some more discussion (the rest of that page is probably good reading too!).

timabbott commented on zerver/tests/test_user_groups.py:
```diff
@@ -361,42 +362,88 @@ def test_update_members_of_user_group(self) -> None:
         result = self.client_post(f"/json/user_groups/{user_group.id}/members", info=params)
         self.assert_json_error(result, f"There is no member '{othello.id}' in this user group")
+
+        # Test user delete itself from user group
```
For example, you could change this test to remove 3 users: (bot, self, deactivated user) and cover all 3 of those corner cases together in a single API request, just adding some extra assertions at the end.

(And of course the previous similar one to add 3 users).

Ujjawal3 replied:
I have added the test as per your comments.

Ujjawal3 commented:
@timabbott I have done the changes as per your comments. Please have a look.

timabbott commented:
It would be extremely helpful if you can explain how you resolved the feedback, e.g. in https://github.com/zulip/zulip/pull/24219/files#diff-49b33fc9770f16be249573ac69270f5541e25f2fc12af6487d071a9d80bc10b4

Ujjawal3 commented:
> It would be extremely helpful if you can explain how you resolved the feedback, e.g. in https://github.com/zulip/zulip/pull/24219/files#diff-49b33fc9770f16be249573ac69270f5541e25f2fc12af6487d071a9d80bc10b4

I have explained how I resolved the feedback. Please have a look.

timabbott commented:
This is much improved, merged, thanks @Ujjawal3!
